src/main/scala/it/pagopa/interop/apigateway/api/impl/GatewayApiServiceImpl.scala

```diff
@@ -0,0 +1,326 @@
+        else validatePurposeIfSubjectIsProducer(subject, purpose)(bearerToken)
+      )
+
+    val result: Future[Purpose] = for {
+      bearerToken          <- getFutureBearer(contexts)
+      subjectUUID          <- getSubFuture(contexts).flatMap(_.toFutureUUID)
+      purposeUUID          <- purposeId.toFutureUUID
+      purpose              <- getPurposeIfAuthorized(subjectUUID, purposeUUID)(bearerToken)
+      actualPurposeVersion <- purpose.toModel.toFuture
+    } yield actualPurposeVersion
+
+    onComplete(result) {
+      case Success(agr) => getPurpose200(agr)
+      case Failure(Forbidden) =>
+        logger.error("The user has no access to the requested purpose {}", purposeId)
+        getPurpose403(problemOf(StatusCodes.Forbidden, Forbidden))
+      case Failure(ex: GenericComponentErrors.ResourceNotFoundError) =>
+        logger.error("Error while getting the requested agreement for purpose {} : {}", purposeId, ex.getMessage)
+        getPurpose404(problemOf(StatusCodes.NotFound, ex))
+      case Failure(e: MissingActivePurposeVersion) =>
+        logger.error("Unable to find an active version of purpose {}", purposeId)
+        getPurpose404(problemOf(StatusCodes.NotFound, e))
+      case Failure(_) => internalServerError
+    }
+  }
+
+  override def getAgreementPurposes(agreementId: String)(implicit
+    contexts: Seq[(String, String)],
+    toEntityMarshallerPurposes: ToEntityMarshaller[Purposes],
+    toEntityMarshallerProblem: ToEntityMarshaller[Problem]
+  ): Route = {
+
+    val result: Future[Purposes] = for {
+      bearerToken   <- getFutureBearer(contexts)
+      subjectUUID   <- getSubFuture(contexts).flatMap(_.toFutureUUID)
+      agreementUUID <- agreementId.toFutureUUID
+      agreement <- agreementManagementService
+        .getAgreementById(agreementUUID)(bearerToken)
+        .ensure(Forbidden)(a => a.consumerId == subjectUUID || a.producerId == subjectUUID)
+      clientPurposes <- purposeManagementService.getPurposes(agreement.eserviceId, agreement.consumerId)(bearerToken)
+      purposes       <- clientPurposes.toModel.toFuture
+    } yield purposes
+
+    onComplete(result) {
+      case Success(agreement) => getAgreementPurposes200(agreement)
+      case Failure(Forbidden) =>
+        logger.error("The user has no access to the requested agreement {}", agreementId)
+        getPurpose403(problemOf(StatusCodes.Forbidden, Forbidden))
+      case Failure(ex: GenericComponentErrors.ResourceNotFoundError) =>
+        logger.error("Error while getting the requested purposes for agreement {} : {}", agreementId, ex.getMessage)
+        getPurpose404(problemOf(StatusCodes.NotFound, ex))
+      case Failure(e: MissingActivePurposesVersions) =>
+        logger.error(
+          "Unable to find active or suspended versions for purposes {} in the agreement {}",
+          e.purposesIds.mkString(", "),
+          agreementId
+        )
+        getAgreementPurposes404(problemOf(StatusCodes.NotFound, e))
```

Comment:
any reason why this is an error and not simply a 200 with an empty list?

Reply:
ping @TonioGela

Reply:
No, not really, maybe it's worth reimplementing it like @galales says